File: _3243_shortest_distance_after_road_addition_queries_i/main.cpp
class Solution {
public:
    std::vector<int> shortestDistanceAfterQueries(int n, std::vector<std::vector<int>>& queries) {
        std::vector<int> result(queries.size());
        std::vector<std::vector<int>> roads(n);
        std::vector<int> dp(n);
        
        // Fill dp
        std::iota(dp.begin(), dp.end(), 0);

        for (int i = 0; i < queries.size(); i++)
        {
            roads[queries[i][1]].push_back(queries[i][0]);
            for (int j = queries[i][1]; j < n; j++)
            {
                dp[j] = min(dp[j], dp[j - 1] + 1);
                for (const int dest : roads[j])
                {
                    dp[j] = min(dp[j], dp[dest] + 1);
                }
            }

            result[i] = dp[n - 1];
        }

        return result;
    }
};

File: _3243_shortest_distance_after_road_addition_queries_i/main.py
class Solution:
    def shortestDistanceAfterQueries(self, n: int, queries: List[List[int]]) -> List[int]:
        graph = defaultdict(list)
        for i in range(0, n - 1):
            graph[i].append(i + 1)

        result = []

        for query in queries:
            c1, c2 = query
            graph[c1].append(c2)
            result.append(self.bfs(graph, n))

        return result


    def bfs(self, graph, n):
        distances = [float("inf") for _ in range(n)]
        distances[0] = 0

        heap = [(0, 0)]

        print(heap)


        while heap:
            distance, node = heapq.heappop(heap)

            for vertex in graph[node]:
                if distances[vertex] > distances[node] + 1:
                    distances[vertex] = distances[node] + 1
                    heapq.heappush(heap, (distances[vertex], vertex))


        return distances[n - 1]
